wl-sim/main/main.cpp

```diff
@@ -0,0 +1,191 @@
+#include <cstdlib>
+#include <time.h>
+#include <cstring>
+
+#include "esp_log.h"
+#include "wl_sim_random.h"
+#include "wl_sim.h"
+#include "WLsim_Flash.h"
+
+static const char *TAG = "wl-sim";
+
+extern uint32_t restarted;
+extern size_t access_count;
+
+typedef size_t (*address_function_t)(size_t);
+typedef size_t (*block_size_function_t)(size_t);
+
+// forward declaration
+int feistel_test();
+
+int main(int argc, char **argv)
+{
+    srand(time(0));
+    esp_err_t result;
+
+    // if single argument 'test', run the mapping correctness test
+    if (argc == 2 && strcmp(argv[1], "test") == 0) {
+        feistel_test();
+        return 0;
+    }
+
+    // otherwise require all args for a simulation run
+    // e.g. wl-sim f z z 10 0
+    // for Feistel enabled, zipf address access and zipf block size with maximum of 10 and 0 per mille chance for restart
+    if (argc != 6) {
+        printf("Need simulation params as arguments:\n\
+\tMAPPING_ALG: f for Feistel, b for base mapping alg\n\
+\tADDRESS_FUNC: z for zipf, c for const\n\
+\tBLOCKS_SIZE_FUNC: z for zipf, c for const\n\
+\tBLOCK_SIZE: N for max erase block size\n\
+\tRESTART_PROB: P for restart probability after every erase [per mille]\n");
+        return ESP_FAIL;
+    }
+
+    // argument parsing
+
+    bool enable_feistel = false;
+    if (strcmp(argv[1], "f") == 0) {
+        enable_feistel = true;
+    } else if (strcmp(argv[1], "b") == 0) {
+        // enable_feistel already false
+    } else {
+        fprintf(stderr, "First argument '%s' invalid, defaulting to base mapping...\n", argv[1]);
+        // enable_feistel already false
+    }
+
+    address_function_t addr_func = &constant;
+    if (strcmp(argv[2], "z") == 0) {
+        addr_func = &zipf;
+    } else if (strcmp(argv[2], "c") == 0) {
+        // addr_func already constant
+    } else {
+        fprintf(stderr, "Second argument '%s' invalid, defaulting to constant address...\n", argv[2]);
+        // addr_func already constant
+    }
+
+    block_size_function_t block_func = &block_constant;
+    if (strcmp(argv[3], "z") == 0) {
+        block_func = &block_zipf;
+    } else if (strcmp(argv[3], "c") == 0) {
+        // block_func already constant
+    } else {
+        fprintf(stderr, "Third argument '%s' invalid, defaulting to constant erase block size...\n", argv[3]);
+        // block_func already constant
+    }
+
+    // last two arguments are numbers
+    char *end = NULL;
+    int erase_block_size = strtol(argv[4], &end, 10);
+    if (*end != '\0') {
+        fprintf(stderr, "Invalid erase block size '%s'!\n", argv[4]);
+        return ESP_FAIL;
```

Comment:
Returning esp_err_t from retval(int) function? It likely compiles but it's bad practice - please, keep the return types consistent, no matter whether in a real code or in a simulator., Applies to all similar cases in the project

Reply:
fixed, thanks